packages/x-data-grid-pro/src/hooks/features/dataSource/useGridDataSourceBasePro.ts

```diff
@@ -304,6 +317,50 @@ export const useGridDataSourceBasePro = <Api extends GridPrivateApiPro>(
     [apiRef],
   );
 
+  const handleGroupedDataUpdate = React.useCallback<GridStrategyProcessor<'dataSourceRowsUpdate'>>(
+    (params) => {
+      if ('error' in params) {
+        apiRef.current.setRows([]);
+        return;
+      }
+
+      const {
+        response,
+        options: { collapseChildren },
+      } = params;
+      if (response.rowCount !== undefined) {
+        apiRef.current.setRowCount(response.rowCount);
+      }
+
+      if (collapseChildren) {
+        apiRef.current.setRows(response.rows);
+      } else {
+        // Remove existing outdated rows before setting the new ones
+        // Create a set of the current root rows
+        const rowsToDelete = new Set(
+          getTreeNodeDescendants(gridRowTreeSelector(apiRef), GRID_ROOT_GROUP_ID, false, true),
```

Comment:
Shouldn't `directChildrenOnly` be `false` here to cleanup all the nested children too? Otherwise we may endup with more rows in state than necessary (stale descendant rows)